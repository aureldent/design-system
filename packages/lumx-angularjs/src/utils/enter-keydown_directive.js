import { ENTER_KEY_CODE } from '@lumx/core/js/constants';

/////////////////////////////

function EnterKeydownDirective() {
    'ngInject';

    function link(scope, el, attrs) {
        el.on('keydown', (evt) => {
            if (evt.which === ENTER_KEY_CODE) {
                scope.$apply(() => {
                    scope.$eval(attrs.lxEnterKeydown, { $event: evt });
                });

                evt.preventDefault();
            }
        });
    }

    return {
        link,
    };
}

/////////////////////////////

angular.module('lumx.utils.enter-keydown').directive('lxEnterKeydown', EnterKeydownDirective);

/////////////////////////////

export { EnterKeydownDirective };
