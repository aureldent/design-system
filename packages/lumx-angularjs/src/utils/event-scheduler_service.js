function EventSchedulerService($document, LxUtilsService) {
    'ngInject';

    const service = this;

    /////////////////////////////
    //                         //
    //    Private attributes   //
    //                         //
    /////////////////////////////

    /**
     * The handlers.
     *
     * @type {Object}
     */
    const _handlers = {};

    /**
     * The schedule.
     *
     * @type {Object}
     */
    const _schedule = {};

    /////////////////////////////
    //                         //
    //    Private functions    //
    //                         //
    /////////////////////////////

    /**
     * Handle en event.
     *
     * @param {Event} evt The event.
     */
    function _handle(evt) {
        let scheduler;

        for (const [key] of Object.entries(_schedule)) {
            if (key.includes(evt.type)) {
                scheduler = _schedule[key];
            }
        }

        if (angular.isDefined(scheduler)) {
            for (let i = 0, len = scheduler.length; i < len; i++) {
                const handler = scheduler[i];

                if (angular.isDefined(handler) && angular.isDefined(handler.cb) && angular.isFunction(handler.cb)) {
                    handler.cb(evt);

                    if (evt.isPropagationStopped()) {
                        break;
                    }
                }
            }
        }
    }

    /////////////////////////////
    //                         //
    //     Public functions    //
    //                         //
    /////////////////////////////

    /**
     * Register en event.
     *
     * @param  {string}   eventName The event name.
     * @param  {Function} cb        The event callback.
     * @return {string}   The event id.
     */
    function register(eventName, cb) {
        const handler = {
            cb,
            eventName,
        };

        const id = LxUtilsService.generateUUID();
        _handlers[id] = handler;

        if (angular.isUndefined(_schedule[eventName])) {
            _schedule[eventName] = [];

            $document.on(eventName, _handle);
        }

        _schedule[eventName].unshift(_handlers[id]);

        return id;
    }

    /**
     * Unregister en event.
     *
     * @param  {string} id The event id.
     * @return {found}  Whether the event has been found or not.
     */
    function unregister(id) {
        let found = false;
        const handler = _handlers[id];

        if (angular.isDefined(handler) && angular.isDefined(_schedule[handler.eventName])) {
            const index = _schedule[handler.eventName].indexOf(handler);

            if (angular.isDefined(index) && index > -1) {
                _schedule[handler.eventName].splice(index, 1);

                delete _handlers[id];
                found = true;
            }

            if (_schedule[handler.eventName].length === 0) {
                delete _schedule[handler.eventName];

                $document.off(handler.eventName, _handle);
            }
        }

        return found;
    }

    /////////////////////////////

    service.register = register;
    service.unregister = unregister;
}

/////////////////////////////

angular.module('lumx.utils.event-scheduler').service('LxEventSchedulerService', EventSchedulerService);

/////////////////////////////

export { EventSchedulerService };
