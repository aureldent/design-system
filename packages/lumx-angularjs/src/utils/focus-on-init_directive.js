import { COMPONENT_PREFIX, MODULE_NAME } from '@lumx/angularjs/constants/common_constants';

/////////////////////////////

function FocusOnInitDirective($timeout) {
    'ngInject';

    function link(scope, el, attrs) {
        if (angular.isDefined(attrs.lumxFocusOnInit) && attrs.lumxFocusOnInit && !scope.$eval(attrs.lumxFocusOnInit)) {
            return;
        }

        $timeout(() => {
            el.focus();
        });
    }

    return {
        link,
    };
}

/////////////////////////////

angular.module(`${MODULE_NAME}.utils.focus-on-init`).directive(`${COMPONENT_PREFIX}FocusOnInit`, FocusOnInitDirective);

/////////////////////////////

export { FocusOnInitDirective };
