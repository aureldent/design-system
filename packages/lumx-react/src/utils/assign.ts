export const assign = (target: any, ...sources: any[]) => {
    sources.forEach((source) => Object.keys(source).forEach((key) => (target[key] = source[key])));
    return target;
};
