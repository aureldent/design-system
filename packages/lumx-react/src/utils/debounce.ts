export function debounce(func: (...args: any) => any, debounceTime: number) {
    let timeout: any;
    return function debouncedFunction(...args: any[]) {
        clearTimeout(timeout);
        timeout = setTimeout(() => func(...args), debounceTime);
    };
}
