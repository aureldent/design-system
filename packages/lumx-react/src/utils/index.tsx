export { getBasicClass, handleBasicClasses, detectSwipe, onEnterPressed, onEscapePressed } from '@lumx/core/utils';

export * from './calculatePopoverPlacement';

export * from './isInViewPort';

export * from './validation';

export * from './partitionMulti';

export { getRootClassName } from './getRootClassName';

export * from './type';
