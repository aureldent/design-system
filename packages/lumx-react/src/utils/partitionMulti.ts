import { Predicate } from './type';
import reduce from 'lodash/reduce';
import concat from 'lodash/concat';
import dropRight from 'lodash/dropRight';
import partition from 'lodash/partition';
import last from 'lodash/last';

/**
 * Similar to lodash `partition` function but working with multiple predicates.
 *
 * @example
 * const isString = (s) => typeof s === 'string'
 * const isNumber = (s) => typeof s === 'number'
 * const [strings, numbers, others] = partitionMulti(['a', 1, 'b', false], [isString, isNumber])
 * //=> [['a', 'b'], [1], [false]]
 *
 * @param  elements array of elements
 * @param  predicates array of predicates to apply on elements
 * @return partitioned elements by the given predicates
 */
function partitionMulti<T>(elements: T[], predicates: Array<Predicate<T>>): T[][] {
    return reduce(
        predicates,
        (partitioned: T[][], predicate: Predicate<T>) =>
            concat(dropRight(partitioned), partition(last(partitioned), predicate)),
        [elements],
    );
}

export { partitionMulti };
