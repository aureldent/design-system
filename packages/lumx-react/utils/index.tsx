export * from './calculatePopoverPlacement';

export * from './isInViewPort';
