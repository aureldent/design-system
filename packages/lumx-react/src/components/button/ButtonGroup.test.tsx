import React, { ReactElement, ReactNode } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';
import mockConsole from 'jest-mock-console';

import { mdiPlus } from '@lumx/icons';
import { Button, Icon, IconButton } from '@lumx/react';
import { ICommonSetup, Wrapper, commonTestsSuite } from '@lumx/react/testing/utils';

import { ButtonGroup, ButtonGroupProps, CLASSNAME } from './ButtonGroup';

/////////////////////////////

/**
 * Define the overriding properties waited by the `setup` function.
 */
type ISetupProps = Partial<ButtonGroupProps>;

/**
 * Defines what is returned by the setup function.
 */
interface ISetup extends ICommonSetup {
    /**
     * The properties of the tested component.
     */
    props: ISetupProps;

    /**
     * The <div> element that is used as a wrapper for the buttons inside of the <ButtonGroup>.
     */
    group: Wrapper;
}

/////////////////////////////

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param props The props to use to override the default props of the component.
 * @param     [shallowRendering=true] Indicates if we want to do a shallow or a full rendering.
 * @return      An object with the props, the component wrapper and some shortcut to some element inside of
 *                       the component.
 */
const setup = ({ ...propsOverrides }: ISetupProps = {}, shallowRendering: boolean = true): ISetup => {
    const props: ButtonGroupProps = {
        children: (
            <>
                <Button>Label</Button>
                <IconButton icon={mdiPlus} />
            </>
        ),
        ...propsOverrides,
    };

    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    const wrapper: Wrapper = renderer(<ButtonGroup {...props} />);

    return {
        group: wrapper.find(`.${CLASSNAME}`),

        props,
        wrapper,
    };
};

describe(`<${ButtonGroup.displayName}>`, (): void => {
    // 1. Test render via snapshot (default state of component).
    describe('Snapshots and structure', (): void => {
        it('should render correctly a group button', (): void => {
            const { group, wrapper } = setup();
            expect(wrapper).toMatchSnapshot();

            expect(group).toExist();
            expect(group).toHaveClassName(CLASSNAME);

            expect(group.children().length).toEqual(2);
        });
    });

    /////////////////////////////

    // 2. Test defaultProps value and important props custom values.
    describe('Props', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // 3. Test events.
    describe('Events', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', (): void => {
        beforeEach((): void => {
            try {
                // If `console.warn` has been mocked at least one, this exists. So disable TS here.
                // @ts-ignore
                global.console.warn.mockRestore();
            } catch (exception) {
                // Nothing to do here.
            }
        });

        it('should fail when no child is given', (): void => {
            expect((): void => {
                setup({ children: null });
            }).toThrowErrorMatchingSnapshot();
        });

        it('should fail when less than 2 children are given', (): void => {
            const children: ReactNode = <Button>Label</Button>;

            expect((): void => {
                setup({ children });
            }).toThrowErrorMatchingSnapshot();
        });

        it(`should fail when anything else than <${Button.displayName}>s or <${IconButton.displayName}> is passed as children`, (): void => {
            mockConsole('debug');

            let children: ReactNode = (
                <>
                    <Icon icon={mdiPlus} />
                    <Icon icon={mdiPlus} />
                </>
            );

            expect((): void => {
                setup({ children });
            }).toThrowErrorMatchingSnapshot();

            /////////////////////////////

            children = (
                <>
                    <span>Label</span>
                    <Icon icon={mdiPlus} />
                </>
            );

            expect((): void => {
                setup({ children });
            }).toThrowErrorMatchingSnapshot();

            /////////////////////////////

            children = (
                <>
                    <span>Label</span>
                    <span>Label 2</span>
                </>
            );

            expect((): void => {
                setup({ children });
            }).toThrowErrorMatchingSnapshot();

            /////////////////////////////

            children = (
                <>
                    <Button>Label</Button>
                    <Icon icon={mdiPlus} />
                </>
            );

            expect((): void => {
                setup({ children });
            }).toThrowErrorMatchingSnapshot();

            /////////////////////////////

            children = (
                <>
                    <Button>Label</Button>
                    <span>Label</span>>
                </>
            );

            expect((): void => {
                setup({ children });
            }).toThrowErrorMatchingSnapshot();

            /////////////////////////////

            children = (
                <>
                    <Icon icon={mdiPlus} />
                    <Button>Label</Button>
                </>
            );

            expect((): void => {
                setup({ children });
            }).toThrowErrorMatchingSnapshot();

            /////////////////////////////

            children = (
                <>
                    <span>Label</span>><Button>Label</Button>
                </>
            );

            expect((): void => {
                setup({ children });
            }).toThrowErrorMatchingSnapshot();
        });

        it('should fail when more than 2 children are given', (): void => {
            const children: ReactNode = (
                <>
                    <Button>Label</Button>
                    <Button>Label 2</Button>
                    <Button>Label 3</Button>
                </>
            );

            expect((): void => {
                setup({ children });
            }).toThrowErrorMatchingSnapshot();
        });
    });

    /////////////////////////////

    // 5. Test state.
    describe('State', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // Common tests suite.
    commonTestsSuite(setup, { className: 'group', prop: 'group' }, { className: CLASSNAME });
});
