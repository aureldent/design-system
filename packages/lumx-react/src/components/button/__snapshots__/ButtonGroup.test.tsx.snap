// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup> Conditions should fail when anything else than <Button>s or <IconButton> is passed as children 1`] = `"You can only have <IconButton> or <Button> children in <ButtonGroup> (got <Icon>)!"`;

exports[`<ButtonGroup> Conditions should fail when anything else than <Button>s or <IconButton> is passed as children 2`] = `"You can only have <IconButton> or <Button> children in <ButtonGroup> (got <span>)!"`;

exports[`<ButtonGroup> Conditions should fail when anything else than <Button>s or <IconButton> is passed as children 3`] = `"You can only have <IconButton> or <Button> children in <ButtonGroup> (got <span>)!"`;

exports[`<ButtonGroup> Conditions should fail when anything else than <Button>s or <IconButton> is passed as children 4`] = `"You can only have <IconButton> or <Button> children in <ButtonGroup> (got <Icon>)!"`;

exports[`<ButtonGroup> Conditions should fail when anything else than <Button>s or <IconButton> is passed as children 5`] = `"<ButtonGroup> must have exactly 2 children (got 3)!"`;

exports[`<ButtonGroup> Conditions should fail when anything else than <Button>s or <IconButton> is passed as children 6`] = `"You can only have <IconButton> or <Button> children in <ButtonGroup> (got <Icon>)!"`;

exports[`<ButtonGroup> Conditions should fail when anything else than <Button>s or <IconButton> is passed as children 7`] = `"<ButtonGroup> must have exactly 2 children (got 3)!"`;

exports[`<ButtonGroup> Conditions should fail when less than 2 children are given 1`] = `"<ButtonGroup> must have exactly 2 children (got 1)!"`;

exports[`<ButtonGroup> Conditions should fail when more than 2 children are given 1`] = `"<ButtonGroup> must have exactly 2 children (got 3)!"`;

exports[`<ButtonGroup> Conditions should fail when no child is given 1`] = `"<ButtonGroup> must have exactly 2 children (got 0)!"`;

exports[`<ButtonGroup> Snapshots and structure should render correctly a group button 1`] = `
<div
  className="lumx-button-group"
>
  <Button>
    Label
  </Button>
  <IconButton
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
  />
</div>
`;
